src/modules/users/users.controller.ts

```diff
@@ -55,36 +126,112 @@ export class UsersController {
+
+    return this.usersService.createUserWithRole(createUserDto, allowedRoles);
   }
 
   @Patch('/admin/:id')
   @Roles(UserRole.SUPER_ADMIN, UserRole.ADMIN)
-  @ApiOperation({ summary: 'Update a user' })
+  @ApiOperation({ summary: 'Update a admin' })
+  @ApiResponse({
+    status: 200,
+    description: 'Admin successfully updated',
+    type: User,
+  })
+  @HttpCode(HttpStatus.OK)
+  updateAdmin(
+    @Param('id', ParseIntPipe) id: number,
+    @Body() updateUserDto: UpdateUserDto,
+  ): Promise<User | null> {
+    const allowedRole: UserRole = UserRole.ADMIN;
+
+    return this.usersService.updateUserById(id, updateUserDto, allowedRole);
+  }
+
+  @Patch('/dispatcher/:id')
+  @Roles(UserRole.ADMIN, UserRole.DISPATCHER)
+  @ApiOperation({ summary: 'Update a dispatcher' })
+  @ApiResponse({
+    status: 200,
+    description: 'Dispatcher successfully updated',
+    type: User,
+  })
+  @HttpCode(HttpStatus.OK)
+  updateDispatcher(
+    @Param('id', ParseIntPipe) id: number,
+    @Body() updateUserDto: UpdateUserDto,
+  ): Promise<User | null> {
+    const allowedRole: UserRole = UserRole.DISPATCHER;
+
+    return this.usersService.updateUserById(id, updateUserDto, allowedRole);
+  }
+
+  @Patch('/driver/:id')
+  @Roles(UserRole.ADMIN, UserRole.DRIVER)
+  @ApiOperation({ summary: 'Update a dispatcher' })
   @ApiResponse({
     status: 200,
-    description: 'User successfully updated',
+    description: 'Dispatcher successfully updated',
     type: User,
   })
   @HttpCode(HttpStatus.OK)
-  update(
+  updateDriver(
     @Param('id', ParseIntPipe) id: number,
     @Body() updateUserDto: UpdateUserDto,
   ): Promise<User | null> {
-    return this.usersService.updateAdmin(id, updateUserDto);
+    const allowedRole: UserRole = UserRole.DRIVER;
+
+    return this.usersService.updateUserById(id, updateUserDto, allowedRole);
   }
 
   @Delete('/admin/:id')
   @Roles(UserRole.SUPER_ADMIN)
-  @ApiOperation({ summary: 'Delete a user' })
+  @ApiOperation({ summary: 'Delete an admin' })
+  @ApiResponse({
+    status: 200,
+    description: 'Admin successfully deleted',
+    type: User,
+  })
+  @HttpCode(HttpStatus.OK)
+  deleteAdmin(@Param('id', ParseIntPipe) id: number): Promise<User> {
+    const allowedRoles: UserRole[] = [UserRole.ADMIN];
+
+    return this.usersService.deleteUser(id, allowedRoles);
+  }
+
+  @Delete('/admin/dispatcher-driver/:id')
+  @Roles(UserRole.ADMIN)
+  @ApiOperation({ summary: 'Delete a dispatcher or driver' })
   @ApiResponse({
     status: 200,
-    description: 'User successfully deleted',
+    description: 'Dispatcher or driver successfully deleted',
     type: User,
   })
   @HttpCode(HttpStatus.OK)
-  delete(@Param('id', ParseIntPipe) id: number): Promise<User> {
-    return this.usersService.deleteAdmin(id);
+  deleteDispatcherOrDriver(
+    @Param('id', ParseIntPipe) id: number,
+  ): Promise<User> {
+    const allowedRoles: UserRole[] = [UserRole.DISPATCHER, UserRole.DRIVER];
```

Comment:
look like its everywhere

Reply:
I created a single function for creating users and other functionality, with allowedRoles for reusability. For example, the function allows super_admin to create only admins, while admins can create dispatchers or drivers. To avoid duplicating the function with only minor conditional differences, I implemented allowedRoles to handle these distinctions.